ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java

```diff
@@ -8815,6 +8815,23 @@ These props are now enabled elsewhere (see commit diffs).  It would be better in
     conf.set(AcidUtils.CONF_ACID_KEY, "true");
     SessionState.get().getConf().set(AcidUtils.CONF_ACID_KEY, "true");
   }
+  
+  private void validateColumnCounts(String dest, Table table, Map<String, String> partitionSpec, 
+      int inColumnCnt, int outColumnCnt) throws SemanticException {
+
+    long staticPartitionColumnCnt = partitionSpec != null ?
+        partitionSpec.values().stream().filter(Objects::nonNull).count() : 0;
+
+    if (inColumnCnt + staticPartitionColumnCnt != outColumnCnt &&
```

Comment:
could we extract `table.getStorageHandler() != null && table.getStorageHandler().alwaysUnpartitioned()` into local var `alwaysUnpartitioned`?

Reply:
````
if (inColumnCnt + staticPartitionColumnCnt != outColumnCnt 
    && (!alwaysUnpartitioned || inColumnCnt != outColumnCnt) ) {
...
}
````

Reply:
Extracted the code into a local var, but about the second part, unfortunately it won't work because I saw that for Iceberg it can be `inColumnCnt + staticPartitionColumnCnt == outColumnCnt` or` inColumnCnt == outColumnCnt` and the suggested expression won't cover the first case.

Reply:
i had the same code at the beginning, but it looks a bit weird if you split it
````
!alwaysUnpartitioned && inColumnCnt + staticPartitionColumnCnt != outColumnCnt 
or
alwaysUnpartitioned && inColumnCnt + staticPartitionColumnCnt != outColumnCnt &&  inColumnCnt != outColumnCnt
````

Reply:
if it could be `inColumnCnt + staticPartitionColumnCnt == outColumnCnt where staticPartitionColumnCnt is > 0 and in other cases it can be  inColumnCnt == outColumnCnt` why do we even need to handle iceberg differently?

if(inColumnCnt + staticPartitionColumnCnt != outColumnCnt)

Reply:
Changed IOW query in Iceberg compactor class and to not include partition keys in the select query and now this change in SemanticAnalyzer is not needed.